File: 74. Search a 2D Matrix.java
// Time Complexity : O(logN)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes, https://leetcode.com/problems/search-a-2d-matrix/
// Any problem you faced while coding this : No


class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if(matrix.length==0) return false;
        int m = matrix.length;
        int n = matrix[0].length;
        int low = 0;
        int high = m*n-1;
        while(low<=high){
            int mid = low+(high-low)/2;
            int r = mid/n;
            int c = mid%n;
            if(matrix[r][c]==target){
                return true;
            }else if(matrix[r][c]<target){
                low = mid+1;
            }else{
                high = mid-1;
            }
        }
        return false;
    }
}

File: 74. Search a 2D Matrix.py
### Perform binary search on the 'rows' to isolate which row has the target (O log m)
###                                  + 
### Perform binary search on the specific row (o log n)
### Time Complexity - O (log (m + n)) 

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:

        ROW = len(matrix)
        COL = len(matrix[0])

        top = 0
        bot = ROW - 1

        ## First we try to find which row contains our target
        ## If number doesn't exist, our top pointer will exceed the back pointer, which means target not found

        while top <= bot:
            row = (top + bot) // 2

            if(target > matrix[row][-1]):
                top = row + 1
            
            elif(target < matrix[row][0]):
                bot = row - 1
            else:
                break
        
        if top > bot:
            return False

        l = 0
        r = COL -1

        ## once we know which row has our target, we perform binary search on it
            
        while(l <= r):
            mid = (l + r) // 2
            if target < matrix[row][mid]:
                r = mid - 1
            elif target > matrix[row][mid]:
                l = mid + 1
            else:
                return True
        
        return False
        


        
